src/providers/ipa/ipa_deskprofile_rules_util.c

```diff
@@ -0,0 +1,796 @@
+/*
+    SSSD
+
+    Authors:
+        Fabiano Fidêncio <fidencio@redhat.com>
+
+    Copyright (C) 2017 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "providers/ipa/ipa_deskprofile_rules_util.h"
+#include "providers/ipa/ipa_deskprofile_private.h"
+#include "providers/ipa/ipa_rules_common.h"
+#include <ctype.h>
+#include <dirent.h>
+#include <sys/types.h>
+
+enum deskprofile_name {
+    RULES_DIR = 0,
+    DOMAIN,
+    USERNAME,
+    PRIORITY,
+    USER,
+    GROUP,
+    HOST,
+    HOSTGROUP,
+    RULE_NAME,
+    EXTENSION,
+    DESKPROFILE_NAME_SENTINEL
+};
+
+static errno_t
+ipa_deskprofile_get_filename_path(TALLOC_CTX *mem_ctx,
+                                  uint16_t config_priority,
+                                  const char *rules_dir,
+                                  const char *domain,
+                                  const char *username,
+                                  const char *priority,
+                                  const char *user_priority,
+                                  const char *group_priority,
+                                  const char *host_priority,
+                                  const char *hostgroup_priority,
+                                  const char *rule_name,
+                                  const char *extension,
+                                  char **_filename_path)
+{
+    TALLOC_CTX *tmp_ctx;
+    static const uint8_t permuts[][DESKPROFILE_NAME_SENTINEL] = {
+        {RULES_DIR, DOMAIN, USERNAME, PRIORITY, USER, GROUP, HOST, HOSTGROUP, RULE_NAME, EXTENSION},
```

Comment:
So I'm normally not a big fan of comments (self-documenting code and all that..) but I think this code should be commented. Do we really need this array rather than some loops?